### PR TITLE
Detail endpoint

### PR DESCRIPTION
**Pull request by amymok**
This PR details out the basic search, suggest, and document functionalities.

## Additions:
- Added logic to take in query parameters for search, suggest, document in views
- Added serializer to better valid the input data
- Added tests
- Added setting for travis and coveralls

## Changes:
- Updated search, suggest, document to take in parameters to put together elasticsearch query

## TODOs:
- ~~Update missing parameters (from Jeff's comment below)~~
- This will not be in this PR:
  - handling of different output format other than json

## Reviews:
 - @AdamZarger @JeffreyMFarley 
- cc @sephcoster @rosskarchner

### CODE REVIEW

File: complaint_search/views.py [hunk 1]
```diff
@@ -11,35 +11,32 @@ def search(request):
     print format
     fixed_qparam = request.query_params
     print fixed_qparam
+    QPARAMS_VARS = ('fmt', 'field', 'size', 'frm', 'sort',
```
Comment (JeffreyMFarley) [inline complaint_search/views.py]: 👏

File: complaint_search/es_interface.py [hunk 2]
```diff
@@ -61,10 +57,8 @@ def _create_and_append_bool_should_clauses(es_field_name, value_list,
 
             filter_list.append({"bool": {"should": f_list}})
 
-
-# def search(fmt="json"):
 def search(fmt="json", field="what_happened", size=10, frm=0,
```
Comment (JeffreyMFarley) [inline complaint_search/es_interface.py]: That's a lot of arguments!

I would think you could use `**kwargs` except for the values you actually want to default -or-

```python
defaults = {...}
composed = defaults.update(**kwargs)
```

File: complaint_search/views.py [hunk 3]
```diff
@@ -50,8 +50,15 @@ def search(request):
 
 @api_view(['GET'])
 def suggest(request):
-    results = es_interface.suggest()
-    return Response(results)
+    QPARAMS_VARS = ("text", "size")
+    data = { k:v for k,v in request.query_params.iteritems() if k in QPARAMS_VARS }
```
Comment (JeffreyMFarley) [inline complaint_search/views.py]: I love python's comprehensions

File: complaint_search/es_interface.py [hunk 4]
```diff
@@ -21,10 +22,125 @@ def get_es():
...
+                    subitem_term = {"terms": {es_subitem_field_name: subitems}}
+                    f_list.append({"and": {"filters": [item_term, subitem_term]}})
+
+            filter_list.append({"bool": {"should": f_list}})
+
+def search(fmt="json", field="what_happened", size=10, frm=0, 
+    sort=None, search_term=None, min_date=None, max_date=None, 
+    company=None, product=None, subproduct=None, issue=None, subissue=None, 
+    state=None, consumer_disputed=None, company_response=None):
+
```
Comment (JeffreyMFarley) [inline complaint_search/es_interface.py]: That's a lot of arguments!

I would think you could use `**kwargs` except for the values you actually want to default -or-

```python
defaults = {...}
composed = defaults.update(**kwargs)
```

File: complaint_search/views.py (hunk 3 above)
Comment (amymok) [inline complaint_search/views.py]: Yesss!

File: complaint_search/es_interface.py (hunk 4 above)
Comment (amymok) [inline complaint_search/es_interface.py]: Yea that's a good idea, I will work on doing the latter, thanks~

File: complaint_search/es_interface.py [hunk 5]
```diff
@@ -21,17 +22,155 @@ def get_es():
...
+# - search_term: the term to be searched
+# - min_date: return only date including and later than this date i.e. 2017-03-02
+# - max_date: return only date before this date, i.e. 2017-04-12 
+# - company: a list of companies you want ["Bank 1", "Bank 2"]
+# - product: 
+# - issue:
+# - state:
+# - consumer_disputed:
+# - company_response:
+def search(**kwargs):
```
Comment (JeffreyMFarley) [inline complaint_search/es_interface.py]: Cross checking with the wireframe, there are a few that are missing:

```
only_narratives
zip_code
timely
company_public_response
consumer_consent_provided
tag  # tag=Older+American&tag=Servicemember
submitted_via
```

Comment (amymok) [inline complaint_search/es_interface.py]: Okay I will take a look

Comment (amymok) [inline complaint_search/es_interface.py]: Added:
```
zip_code
timely
company_public_response
```

The only fields are either missing from the mapping or I am not sure which mapping they map to.  Will talk to @AdamZarger about those.

**Comment by sephcoster** [comment]
👍

**Pull request merged**